internal/controller/apiextensions/composite/composition_functions.go

```diff
@@ -825,3 +858,86 @@ func UpdateResourceRefs(xr resource.ComposedResourcesReferencer, desired Compose
 
 	xr.SetResourceReferences(refs)
 }
+
+// A PatchingManagedFieldsUpgrader uses a JSON patch to upgrade an object's
+// managed fields from client-side to server-side apply. The upgrade is a no-op
+// if the object does not need upgrading.
+type PatchingManagedFieldsUpgrader struct {
+	client client.Writer
+}
+
+// NewPatchingManagedFieldsUpgrader returns a ManagedFieldsUpgrader that uses a
+// JSON patch to upgrade and object's managed fields from client-side to
+// server-side apply.
+func NewPatchingManagedFieldsUpgrader(w client.Writer) *PatchingManagedFieldsUpgrader {
+	return &PatchingManagedFieldsUpgrader{client: w}
+}
+
+// Upgrade the supplied composed object's field managers from client-side to server-side
+// apply.
+//
+// This is a multi-step process.
+//
+// Step 1: All fields are owned by manager 'crossplane' operation 'Update'. This
+// represents all fields set by the XR controller up to this point.
+//
+// Step 2: Upgrade is called for the first time. We clear all field managers.
+//
+// Step 3: The XR controller server-side applies its fully specified intent
+// as field manager with prefix 'apiextensions.crossplane.io/composed/'. This becomes the
+// manager of all the fields that are part of the XR controller's fully
+// specified intent. All existing fields the XR ontroller didn't specify
+// become owned by a special manager - 'before-first-apply', operation 'Update'.
+//
+// Step 4: Upgrade is called for the second time. It deletes the
+// 'before-first-apply' field manager entry. Only the XR composed field manager
+// remains.
+func (u *PatchingManagedFieldsUpgrader) Upgrade(ctx context.Context, obj client.Object) error {
+	// The composed resource doesn't exist, nothing to upgrade.
+	if !meta.WasCreated(obj) {
+		return nil
+	}
+
+	foundSSA := false
+	foundBFA := false
+	idxBFA := -1
+
+	for i, e := range obj.GetManagedFields() {
+		if strings.HasPrefix(e.Manager, FieldOwnerComposedPrefix) {
+			foundSSA = true
+		}
+		if e.Manager == "before-first-apply" {
+			foundBFA = true
+			idxBFA = i
```

Comment:
is it possible to have multiple managers of type `before-first-apply`? if it is possible (i genuinely don't know), do we need to keep track of each of their index?

Reply:
no, i don't think a K8s resource `metadata.managedFields`  could end up with multiple managedField entries with manager as `before-first-apply`.